src/app/identity.rs

```diff
@@ -0,0 +1,220 @@
+use futures::{Async, Future, Poll, Stream};
+use futures_watch::{Store, Watch};
+use std::sync::Arc;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use tokio_timer::{clock, Delay};
+use tower_grpc::{self as grpc, generic::client::GrpcService, BoxBody};
+
+use api::identity as api;
+use never::Never;
+
+use identity;
+pub use identity::{Crt, CrtKey, InvalidName, Key, Name, TokenSource, TrustAnchors, CSR};
+use transport::tls;
+
+#[derive(Clone, Debug)]
+pub struct Config {
+    pub svc: super::control::ControlAddr,
+    pub trust_anchors: TrustAnchors,
+    pub key: Key,
+    pub csr: CSR,
+    pub token: TokenSource,
+    pub local_name: Name,
+    pub min_refresh: Duration,
+    pub max_refresh: Duration,
+}
+
+#[derive(Clone, Debug)]
+pub struct Local {
+    trust_anchors: TrustAnchors,
+    name: Name,
+    crt_key: Watch<Option<CrtKey>>,
+}
+
+pub type CrtKeyStore = Store<Option<CrtKey>>;
+
+/// Drives updates.
+pub struct Daemon<T>
+where
+    T: GrpcService<BoxBody>,
+    T::ResponseBody: grpc::Body,
+{
+    config: Config,
+    client: api::client::Identity<T>,
+    crt_key: Store<Option<CrtKey>>,
+    expiry: SystemTime,
+    inner: Inner<T>,
+}
+
+enum Inner<T>
+where
+    T: GrpcService<BoxBody>,
+    T::ResponseBody: grpc::Body,
+{
+    Waiting(Delay),
+    ShouldRefresh,
+    Pending(grpc::client::unary::ResponseFuture<api::CertifyResponse, T::Future, T::ResponseBody>),
+}
+
+// === impl Config ===
+
+impl Config {
+    fn refresh(&self, expiry: SystemTime) -> Delay {
+        let now = clock::now();
+
+        let refresh = match expiry
+            .duration_since(SystemTime::now())
+            .ok()
+            .map(|d| d * 7 / 10)
```

Comment:
Mind explaining what this math wizardry is doing? Or else we'll forget in a couple months.

Reply:
Ah, yes. The intent is to get to ~70% of the expiration time. I'll add a comment.

Is there a better way to accomplish that math?

Reply:
Oh sorry, yes I recognized it as 70%. I just didn't get the *why*.

I guess when we get a token, it has an expiration point. And this is saying to start checking for a new token once we're within 30% of it expiring?

Reply:
+1 for a comment